src/v2.rs

```diff
@@ -128,29 +123,42 @@ impl Party {
             return Err(DkgError::BadPublicShares(bad_ids));
         }
 
-        let mut not_enough_shares = Vec::new();
-        for key_id in &self.key_ids {
-            if shares[key_id].len() != comms.len() {
-                warn!(
-                    "key_id {} has {} private shares not {}",
-                    key_id,
-                    shares[key_id].len(),
-                    comms.len()
-                );
-                not_enough_shares.push(*key_id);
+        let mut missing_shares = Vec::new();
+        for dst_key_id in &self.key_ids {
+            for src_key_id in public_shares.keys() {
+                match private_shares.get(dst_key_id) {
+                    Some(shares) => {
+                        if shares.get(src_key_id).is_none() {
+                            missing_shares.push((*dst_key_id, *src_key_id));
+                        }
+                    }
+                    None => {
+                        missing_shares.push((*dst_key_id, *src_key_id));
+                    }
+                }
             }
         }
-        if !not_enough_shares.is_empty() {
-            return Err(DkgError::NotEnoughShares(not_enough_shares));
+        if !missing_shares.is_empty() {
+            return Err(DkgError::MissingPrivateShares(missing_shares));
         }
 
         let mut bad_shares = Vec::new();
         for key_id in &self.key_ids {
-            for (sender, s) in &shares[key_id] {
-                let comm = &comms[sender];
-                if s * G != compute::poly(&compute::id(*key_id), &comm.poly)? {
-                    bad_shares.push(*sender);
+            if let Some(shares) = private_shares.get(key_id) {
+                for (sender, s) in shares {
+                    if let Some(comm) = public_shares.get(sender) {
+                        if s * G != compute::poly(&compute::id(*key_id), &comm.poly)? {
+                            bad_shares.push(*sender);
+                        }
+                    } else {
+                        warn!("unable to check private share from {}: no corresponding public share, even though we checked for it above", sender);
+                    }
                 }
+            } else {
+                warn!(
+                    "no private shares for key_id {}, even though we checked for it above",
+                    key_id
+                );
```

Comment:
Usually you would write these logs like so:
```rust
warn!(%sender, "unable to check private share: no corresponding public share");
// ...
warn!(%key_id, "no private shares for key_id");
```
Also, the original log messages gave me the impression that this loop and the one above can be consolidated into one. It isn't clear how to do so in a nice way, so I having the two loops is fine.

Reply:
We could certainly do a whole logging refactoring for this and other style issues. 

Yes, we certainly could have done the key share folding along with the error checking. We do that for the group key.

Reply:
https://github.com/Trust-Machines/wsts/issues/99

Reply:
Oh, whoops I thought this crate used tracing.